tests/build/run_cython_not_in_cwd.srctree

```diff
@@ -0,0 +1,56 @@
+
+PYTHON check_paths_from_cython.py
+
+######## src/foo/__init__.py ########
+
+# foo is a package and foo.bar is the module to build
+
+######## src/foo/bar.pyx ########
+
+def func(x):
+    return 2*x
+
+######## build/src/foo/placeholder.txt ########
+
+# Build outputs go in the build directory and cython is run from here.
+
+######## check_paths_from_cython.py ########
+
+import sys
+import subprocess
+
+python = sys.executable
+
+def cython(cmdstr, cwd=None):
+    argv = [sys.executable, '-m', 'cython'] + cmdstr.split()
+    subprocess.run(argv, check=True, cwd=cwd)
+
+def grep(fname, pattern):
+    with open(fname) as fin:
+        return [line for line in fin if pattern in line]
+
+cython('-3 src/foo/bar.pyx -o build/src/foo/bar1.pyx.c')
+cython('-3 ../src/foo/bar.pyx -o src/foo/bar2.pyx.c', cwd='build')
+
+res1 = grep('build/src/foo/bar1.pyx.c', 'bar.pyx')
+res2 = grep('build/src/foo/bar2.pyx.c', 'bar.pyx')
+
+assert res1 == [
+    # Running cython on a file in CWD uses some paths relative to CWD.
+    '  "src/foo/bar.pyx",\n',
+    'static const char __pyx_k_src_foo_bar_pyx[] = "src/foo/bar.pyx";\n',
+    '/* "foo/bar.pyx":2\n',
+    '  /* "foo/bar.pyx":3\n',
+    '  /* "foo/bar.pyx":2\n',
+    '  /* "foo/bar.pyx":2\n'
+], '\n' + ''.join(res1)
```

Comment:
Putting a complete list of occurrences into a test may let the test break if we make non-trivial changes to the code generation elsewhere. I'd rather extract the paths, deduplicate them, and make sure that there are no other path variants in there.

You could use a regex like `r"pkg[/\\]mod[.]pyx"` to make sure it hits any path occurrence, Windows or Posix.

Reply:
If you run this on test on master then for the out of cwd case the output is like:
```
  "bar.pyx",
static const char __pyx_k_home_oscar_current_active_cytho[] = "/home/oscar/current/active/cython/TEST_TMP/build/run_cython_not_in_cwd2/src/foo/bar.pyx";
/* "foo/bar.pyx":2
  /* "foo/bar.pyx":3
  /* "foo/bar.pyx":2
  /* "foo/bar.pyx":2
```
If you also remove the `__init__.py` then the output is like this in both cases:
```
  "src/foo/bar.pyx",
static const char __pyx_k_src_foo_bar_pyx[] = "src/foo/bar.pyx";
/* "bar.pyx":2
  /* "bar.pyx":3
  /* "bar.pyx":2
  /* "bar.pyx":2
```
We want to extract the full path so that we can confirm that it is not an absolute path. We also need to pick up the lone basename bar.pyx as well though because we don't want that either.

I'll adjusted the test to extract a set of paths but I prefer it as it is now because it shows you directly what you would see if you grepped the .c file yourself.

I realise this kind of test is fragile but there need to be some tests that are fragile so that you can see if something like this changes. The new test is more permissive of changes to exactly which paths are output in different places.